File: User/ugcp_svc/Content/ProcessOutput/Commands/CX52.cpp
#include "../../framework.h"
#include "CX52.h"
#include "../../X52/USBX52Write.h"

/// <summary>
/// Comandos del X52
/// </summary>
/// <returns><para>TRUE: procesado y continuar</para><para>FALSE: no procesado</para></returns>
bool CX52::Process(CEventPacket* queue)
{
	bool processed = true;
	PEV_COMMAND command = queue->GetCommandQueue()->front();

	switch (command->Type.Get())
	{
		case CommandType::X52MfdLight:
		{
			UCHAR params = command->Basic.Data1;
			CX52Write::Get().Light_MFD(params);
			break;
		}
		case CommandType::X52Light:
		{
			UCHAR params = command->Basic.Data1;
			CX52Write::Get().Light_Global(params);
			break;
		}
		case CommandType::X52InfoLight:
		{
			UCHAR params = command->Basic.Data1;
			CX52Write::Get().Light_Info(params);
			break;
		}
		case CommandType::X52MfdPinkie:
		{
			UCHAR params = command->Basic.Data1;
			CX52Write::Get().Set_Pinkie(params);
			break;
		}
		case CommandType::X52MfdTextIni:
		{
			std::uint8_t text[17]{};
			UCHAR size = 1;

			text[0] = command->Basic.Data1; //line
			while (queue->GetCommandQueue()->size() != 1)
			{
				std::deque<PEV_COMMAND>::iterator pos = queue->GetCommandQueue()->begin();
				PEV_COMMAND comTxt = *(++pos);
				if (comTxt->Type == CommandType::X52MfdTextEnd)
				{
					delete comTxt;
					queue->GetCommandQueue()->erase(pos);
					break;
				}
				if (size == 17)
				{
					throw new std::exception("Error text buffer");
				}
				text[size++] = comTxt->Basic.Data1;
				delete comTxt;
				queue->GetCommandQueue()->erase(pos);
			}
			CX52Write::Get().Set_Text(std::span<const std::uint8_t>(text).subspan(0, size));
			break;
		}
		case CommandType::X52MfdHour:
		{
			std::array<std::uint8_t, 3> params{};
			params[0] = command->Basic.Data1;

			std::deque<PEV_COMMAND>::iterator pos = queue->GetCommandQueue()->begin();
			params[1] = (*(++pos))->Basic.Data1;
			delete (*pos);
			queue->GetCommandQueue()->erase(pos);

			pos = queue->GetCommandQueue()->begin();
			params[2] = (*(++pos))->Basic.Data1;
			delete (*pos);
			queue->GetCommandQueue()->erase(pos);

			CX52Write::Get().Set_Hour(params);
			break;
		}
		case CommandType::X52MfdHour24:
		{
			std::array<std::uint8_t, 3> params{};
			params[0] = command->Basic.Data1;

			std::deque<PEV_COMMAND>::iterator pos = queue->GetCommandQueue()->begin();
			params[1] = (*(++pos))->Basic.Data1;
			delete (*pos);
			queue->GetCommandQueue()->erase(pos);

			pos = queue->GetCommandQueue()->begin();
			params[2] = (*(++pos))->Basic.Data1;
			delete (*pos);
			queue->GetCommandQueue()->erase(pos);

			CX52Write::Get().Set_Hour24(params);
			break;
		}
		case CommandType::MfdDate:
		{
			std::array<std::uint8_t, 2> params{};
			params[0] = command->Basic.Data1;

			std::deque<PEV_COMMAND>::iterator pos = queue->GetCommandQueue()->begin();
			params[1] = (*(++pos))->Basic.Data1;
			delete (*pos);
			queue->GetCommandQueue()->erase(pos);

			CX52Write::Get().Set_Date(params);
			break;
		}
		default:
			processed = false;
			break;
	}

	return processed;
}

File: User/ugcp_svc/ProcessOutput/Commands/CX52.cs
using System;
using ugcp_svc.EventQueue;

namespace ugcp_svc.ProcessOutput.Commands
{
    static class CX52
    {
        /// <summary>
        /// Comandos del X52
        /// </summary>
        /// <returns><para>TRUE: procesado y continuar</para><para>FALSE: no procesado</para></returns>
        public static bool Process(CEventPacket queue)
        {
            bool processed = true;
            ref EV_COMMAND command = ref System.Runtime.InteropServices.CollectionsMarshal.AsSpan(queue)[0];

            switch (command.Type.Get())
            {
                case CommandType.X52MfdLight:
                    {
                        byte _params = command.Data.Basic.Data1;
                        X52.CX52Write.Get()?.Light_MFD(_params);
                        break;
                    }
                case CommandType.X52Light:
                    {
                        byte _params = command.Data.Basic.Data1;
                        X52.CX52Write.Get()?.Light_Global(_params);
                        break;
                    }
                case CommandType.X52InfoLight:
                    {
                        byte _params = command.Data.Basic.Data1;
                        X52.CX52Write.Get()?.Light_Info(_params);
                        break;
                    }
                case CommandType.X52MfdPinkie:
                    {
                        byte _params = command.Data.Basic.Data1;
                        X52.CX52Write.Get()?.Set_Pinkie(_params);
                        break;
                    }
                case CommandType.X52MfdTextIni:
                    {
                        Span<byte> text = new byte[17];
                        byte size = 1;

                        text[0] = command.Data.Basic.Data1; //line
                        while (queue.Count != 1)
                        {
                            if (queue[1].Type == CommandType.X52MfdTextEnd)
                            {
                                queue.RemoveAt(1);
                                break;
                            }
                            if (size == 17)
                            {
                                throw new Exception("Error text buffer");
                            }
                            text[size++] = queue[1].Data.Basic.Data1;
                            queue.RemoveAt(1);
                        }
                        X52.CX52Write.Get()?.Set_Text(text.Slice(0, size));
                        break;
                    }
                case CommandType.X52MfdHour:
                    {
                        Span<byte> _params = new byte[3];
                        _params[0] = command.Data.Basic.Data1;

                        _params[1] = queue[1].Data.Basic.Data1;
                        queue.RemoveAt(1);

                        _params[2] = queue[1].Data.Basic.Data1;
                        queue.RemoveAt(1);

                        X52.CX52Write.Get()?.Set_Hour(_params);
                        break;
                    }
                case CommandType.X52MfdHour24:
                    {
                        Span<byte> _params = new byte[3];
                        _params[0] = command.Data.Basic.Data1;

                        _params[1] = queue[1].Data.Basic.Data1;
                        queue.RemoveAt(1);

                        _params[2] = queue[1].Data.Basic.Data1;
                        queue.RemoveAt(1);

                        X52.CX52Write.Get()?.Set_Hour24(_params);
                        break;
                    }
                case CommandType.MfdDate:
                    {
                        Span<byte> _params = new byte[2];
                        _params[0] = command.Data.Basic.Data1;

                        _params[1] = queue[1].Data.Basic.Data1;
                        queue.RemoveAt(1);

                        X52.CX52Write.Get()?.Set_Date(_params);
                        break;
                    }
                default:
                    processed = false;
                    break;
            }

            return processed;
        }

    }
}
